test/cpp/api/inference_mode.cpp

```diff
@@ -0,0 +1,391 @@
+  }
+}
+
+TEST(InferenceModeTest, TestInferenceTensorInInferenceModeViewOp) {
+  c10::InferenceMode guard;
+  for (bool requires_grad : {true, false}) {
+    torch::Tensor c = torch::ones({1, 2, 3}).set_requires_grad(requires_grad);
+
+    torch::Tensor view_out = view_op(c);  // go through kernels: CPU
+    ASSERT_TRUE(is_inference_tensor(view_out));
+    // Note this is different from NoGradMode but makes sense.
+    ASSERT_FALSE(view_out.requires_grad());
+  }
+}
+
+TEST(InferenceModeTest, TestInferenceTensorInNormalModeFunctionalOp) {
+  torch::Tensor inference_tensor;
+  for (bool requires_grad: {true, false}) {
+    {
+      InferenceMode guard;
+      inference_tensor = torch::ones({1, 2, 3}).set_requires_grad(requires_grad);
+    }
+    // Functional ops on inference tensors might run slower outside InferenceMode than inside.
+    // But it's fine that we don't care about perf of this case that much.
+    //
+    // An alternative behavior we perfer but didn't implement is throwing an error by forcing
+    // this op go through VariableType kernel and hit the assert_no_inference_tensor check.
+    // But to do that we'll have to add c10::autograd_dispatch_keyset to the globally enabled set,
+    // but doing that might accidentally call autograd kernel from a backend that doesn't match tensor input.
+    // Thus we allow functional ops run without throwing an error.
+    torch::Tensor tmp = functional_op(inference_tensor); // go through kernels: InplaceOrView(fallthrough), CPU
+    ASSERT_FALSE(is_inference_tensor(tmp));
+    ASSERT_FALSE(tmp.requires_grad());
+  }
+}
+
+TEST(InferenceModeTest, TestInferenceTensorInNormalModeInplaceOp) {
+  torch::Tensor inference_tensor;
+  for (bool requires_grad: {true, false}) {
+    {
+      InferenceMode guard;
+      inference_tensor = torch::ones({1, 2, 3});
+    }
+    ASSERT_THROWS_WITH(inplace_op(inference_tensor), // go through kernels: InplaceOrView(ERROR!), CPU
+      "inplace/view ops on inference tensor outside InferenceMode");
+  }
+}
+
+TEST(InferenceModeTest, TestInferenceTensorInNormalModeViewOp) {
+  torch::Tensor inference_tensor;
+  for (bool requires_grad: {true, false}) {
+    {
+      InferenceMode guard;
+      inference_tensor = torch::ones({1, 2, 3});
+    }
+    ASSERT_THROWS_WITH(view_op(inference_tensor), // go through kernels: InplaceOrView(ERROR!), CPU
+      "inplace/view ops on inference tensor outside InferenceMode")
+  }
+}
+
+TEST(InferenceModeTest, TestNormalTensorInplaceOpInInferenceMode) {
+  for (bool requires_grad: {true, false}) {
+    torch::Tensor s = torch::ones({1, 2, 3}).set_requires_grad(requires_grad);
+    torch::Tensor a = s + 2;
+
+    {
+      c10::InferenceMode guard;
+
+      inplace_op(a);  // go through kernels: InplaceOrView, CPU
+      ASSERT_FALSE(is_inference_tensor(a));
+      ASSERT_EQ(a.requires_grad(), requires_grad);
+
+      // inplace -> inplace
+      inplace_op(a);  // go through kernels: InplaceOrView, CPU
```

Comment:
Why do you do this twice?

Reply:
no particular reason, just thinking that we might want to have a test on chaining ops. ;)